docs/design/loadpath.md

```diff
@@ -0,0 +1,284 @@
+
+Given that a lot of the software using eBPF is also Cloud-Native, we have the
+following additional benefits:
+
+1. A Cloud-Native, declarative way to load eBPF programs and have them
+   orchestrated by Kubernetes. This is backed by packaging eBPF programs as
+   container images.
+1. For daemonsets/deployments that use these eBPF programs to not require
+   privileged containers, host mounts or any capabilities.
+
+These benefits however are proving to be insufficient to sway prospective users
+away from loading their own eBPF programs for 2 reasons:
+
+1. They are an established project and has code that works and don't see a
+   reason to change it
+1. They believe that separating the user-space code and kernel-space code is
+   a bad idea, citing the following reasons:
+   - The versions of the kernel and user-space must be tightly coupled.
+   - They are wedded to BPF Skeletons (bpftool generated C code) since they
+     require complex .bss/.rodata section manipulation before loading, and they
+     argue that this is too complex to model via an API.
+
+This design document explores a way to address these concerns, while still
+providing the benefits of the current bpfman design.
+
+## Design
+
+### Requirements
+
+We would like for bpfman to still be valuable to users who don't want to use
+our opinionated loading pattern. To address the concerns raised above, we
+must allow users to package and load their eBPF bytecode as they see fit.
+However we are left with the following gaps:
+
+1. Signing of eBPF programs - previously enabled via cosign.
+1. Removing CAP_BPF from long running processes or daemonsets/deployments
+   that use eBPF programs.
+1. Simplified lifecycle management of eBPF programs through pinning.
+
+We can immediately discard "simplified lifecycle management" as a requirement
+since that becomes the responsibility of the user. However, we will evaluate
+potential solutions to the other 2 requirements.
+
+### Signing
+
+Therefore, in order for bpfman to NOT be in the load path, we need to find a way
+to preserve the security benefits of our current design.
+
+#### fsverity and LSM Gatekeeper
+
+[BPF Signing using fsverity and LSM Gatekeepers] is a proposal that was put
+forward by Lorenz Bauer. It proposes a way to sign programs that contain
+embedded eBPF bytecode (or eBPF bytecode itself), the kernel will then verify
+the signature when the program is loaded. LSM hooks can the be used to further
+apply policy to the program that's being loaded.
+
+This prevents a bad actor from modifying the eBPF bytecode in a program, since
+the signature will no longer be valid.
+
+This proposal is still a work in progress, but it's a promising development.
+In particular, the integration with the IMA subsystem is very interesting as it
+opens up the possibility of remote attestations using Keylime. Not only that,
+the more I learn about IMAs capabilities, the more I think it's a good fit for
+eBPF.
```

Comment:
But AFAICT, the proposal is not to sign the actual BPF bytecode, it's just to verify that the binary trying to load a BPF program is trusted to do so? Which is basically the same that BPF token is doing, the fsverity thing is just a way of establishing who gets that trust. There is no restriction on *which* bytecode is being loaded into the kernel, so the application still gets the same blank check as if does with bpf token...

Reply:
Ah yes I'd assumed that the signed binary contained bytecode embedded in an ELF section - which is the most common way for doing it. It could load the bytecode from disk, but unless you also protect that file with fs-verity you could be loading anything. I'll add some more wording here to clarify.

Reply:
Well, in either case, if you can modify the binary you can also just modify it to completely rewrite the bytecode before loading it into the kernel, so whether the BPF bytecode itself is covered by the signature doesn't really matter as it's only the integrity of the userspace binary that is being checked, not the bytecode itself...